x/stake/handler_test.go

```diff
@@ -635,6 +633,41 @@ func TestUnbondingPeriod(t *testing.T) {
 	require.False(t, found, "should have unbonded")
 }
 
+func TestUnbondingFromUnbondingValidator(t *testing.T) {
+	ctx, _, keeper := keep.CreateTestInput(t, false, 1000)
+	validatorAddr, delegatorAddr := sdk.ValAddress(keep.Addrs[0]), keep.Addrs[1]
+
+	// create the validator
+	msgCreateValidator := newTestMsgCreateValidator(validatorAddr, keep.PKs[0], 10)
+	got := handleMsgCreateValidator(ctx, msgCreateValidator, keeper)
+	require.True(t, got.IsOK(), "expected no error on runMsgCreateValidator")
+
+	// bond a delegator
+	msgDelegate := newTestMsgDelegate(delegatorAddr, validatorAddr, 10)
+	got = handleMsgDelegate(ctx, msgDelegate, keeper)
+	require.True(t, got.IsOK(), "expected ok, got %v", got)
+
+	// unbond the validators bond portion
+	msgBeginUnbondingValidator := NewMsgBeginUnbonding(sdk.AccAddress(validatorAddr), validatorAddr, sdk.NewDec(10))
+	got = handleMsgBeginUnbonding(ctx, msgBeginUnbondingValidator, keeper)
+	require.True(t, got.IsOK(), "expected no error")
+	var finishTime time.Time
+	types.MsgCdc.MustUnmarshalBinary(got.Data, &finishTime)
+	ctx = ctx.WithBlockTime(finishTime.Add(time.Second * -1))
+	EndBlocker(ctx, keeper)
+
+	// test that the delegator can still withdraw their bonds
+	msgBeginUnbondingDelegator := NewMsgBeginUnbonding(delegatorAddr, validatorAddr, sdk.NewDec(10))
+	got = handleMsgBeginUnbonding(ctx, msgBeginUnbondingDelegator, keeper)
+	require.True(t, got.IsOK(), "expected no error")
+	types.MsgCdc.MustUnmarshalBinary(got.Data, &finishTime)
+	ctx = ctx.WithBlockTime(finishTime)
+	EndBlocker(ctx, keeper)
+
+	_, found := keeper.GetUnbondingDelegation(ctx, delegatorAddr, validatorAddr)
+	require.False(t, found, "should be removed from state")
```

Comment:
Okay I _think_ you're test in on the right track and maybe accomplishes what I was hoping to demonstrate - but it's difficult to understand this test - please add more comments describing the time which each of these transactions are taking place on. 

Just to reiterate - what I wanted to verify is that, if a delegator undelegates from a validator which has already gotten kicked (and is therefor unbonding) - then that delegator will enter the bonding queue IN THE MIDDLE of the queue (not at the end) because it requires less unbonding time due to the fact that the state of the tokens has already been unbonding for some time. From here I'd like to see that the unbond completes at the appropriate time as expected (even though it entered in the middle of the queue) - does this make sense?

Reply:
Is this functionality already implemented?  Where an unbonding from a jailed validator counts the time the validator has been unbonding?  Because I tested it, and when I send a `MsgBeginUnbonding` to a delegator delegated to a jailed validator, it sets the `MinTime` based on the unbonding period starting from the Msg, not from when the Validator started unbonding.

Reply:
Added

Reply:
:)